CODEWALK.md

```diff
@@ -346,6 +346,118 @@ essentially who is storing what data, for what fee and which sectors have been s
+organized according to the _distance metric proposed_ in the [Kademlia
+paper](http://www.scs.stanford.edu/~dm/home/papers/kpos.pdf).
+
+To populate our routing table, and to keep it healthy over time, we traverse
+the network periodically (and once on startup) by performing random walks. We
+kickstart this process by seeding our routing table with a set of trusted
+bootstrap nodes operated by the Filecoin project. This approach will likely
+evolve in the future.
+
+On each round, we generate a random peer ID and query our known DHT peers by
+sending them `FindPeer(id)` messages. Each node returns a list of peers they
+believe to be closer to the search target (based on the Kademlia distance
+metric), along with their addresses.
+
+As replies come in, we connect to new incoming nodes we've discovered, and we
+repeat this process iteratively, until either:
+
+a. we find the peer; highly unlikely scenario, as the chance is 1/2^256 with
+   our peer ID length; or
+b. we yield because a timer runs out; the most probable scenario. This is a
+   healthy timeout to timebox discovery iterations.
+
+Throughout this process, the libp2p stack generates `CONNECTED` events for
+every new connection we establish to a peer, for which `go-filecoin` registers
+a callback that triggers the Filecoin `HELLO` protocol negotiation. If the
+other party responds positively, chain sync with that peer begins.
+
+### NAT topology determination
+
+In a P2P world, one of the main challenges is achieving connectivity when one,
+or both, of the parties is/are behind a Network Address Translation (NAT)
+devices, such as routers. NATs are commonplace in different network types: 
+residential, companies, public hotspots, universities, etc. Their purpose is 
+to enable devices in a private network to share a common public IP address in 
+the global, scarce IPv4 space.
+
+Outbound connections behind a NAT tend to be frictionless, but inbound
+connections are problematic. For a primer on this topic, we recommend reading
+[Peer-to-Peer Communication Across Network Address
+Translators](https://pdos.csail.mit.edu/papers/p2pnat.pdf).
+
+To enable a pair of Filecoin peers behind NATs to communicate with one
+another, we use libp2p relay services: nodes deployed across the globe that
+pipe encrypted traffic between any two Filecoin peers, akin to [TURN-like
+servers](https://en.wikipedia.org/wiki/Traversal_Using_Relays_around_NAT).
+
+Relay capability is activated by sensing our NAT status, once at start (may
+take a few minutes) and continuously at runtime. Two protocols are involved
+here:
+
+* **Identify** protocol: for every peer we connect to, we use the *Identify*
+  protocol to learn the address that peer observes from us (source
+  ip_address:port). We compile all our observed addresses as we learn them.
+* **AutoNAT** protocol: for every peer we connect to that supports the AutoNAT
+  service, we request a _dialback_ to our observed addresses. We wait 42
```

Comment:
> 42

:+1:

Reply:
😂

Reply:
Yeah, checking with the author.

Reply:
Removed the number; it's a bit more complex than that. Filed an issue for speeding up the NAT assertion process: https://github.com/libp2p/go-libp2p-autonat/issues/11